src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -0,0 +1,518 @@
+                self.logger.info(f"Namespace {namespace} already exists")
+                # ... but we know that we don't have a lab for the user,
+                # because we got this far.  So there's a stranded namespace,
+                # and we should delete it and recreate it.
+                #
+                # The spec actually calls for us to delete the lab and then the
+                # namespace, but let's just remove the namespace, which should
+                # also clean up all its contents.
+                await self.delete_namespace(namespace)
+                # Just try again, and return *that* one's return value.
+                return await self.create_user_namespace(namespace)
+            else:
+                estr = f"Failed to create namespace {namespace}: {e}"
+                self.logger.exception(estr)
+                raise NSCreationError(estr)
+        # We don't want to give control back until the namespace exists,
+        # because the next step will be creating objects in that namespace.
+        await self._wait_for_operation(
+            namespace=namespace,
+            item="namespace",
+            operation=ObjectOperation.CREATION,
+            interval=0.2,
+        )
+
+    async def _k8s_create_namespace(self, ns_name: str) -> None:
+        await asyncio.wait_for(
+            self.api.create_namespace(
+                V1Namespace(metadata=self.get_std_metadata(name=ns_name))
+            ),
+            self.timeout,
+        )
+
+    async def create_secrets(
+        self,
+        secret_list: List[LabSecret],
+        username: str,
+        token: str,
+        source_ns: str,
+        target_ns: str,
+    ) -> None:
+        pull_secrets = [
+            x for x in secret_list if x.secret_name == "pull_secret"
+        ]
+        secrets = [x for x in secret_list if x.secret_name != "pull_secret"]
+        data = await self.merge_controller_secrets(
+            secret_list=secrets, token=token, source_ns=source_ns
+        )
+        await self.create_secret(
+            name=f"nb-{username}",
+            namespace=target_ns,
+            data=data,
+        )
+        if pull_secrets:
+            pull_secret = await self.copy_pull_secret(source_ns=source_ns)
+            await self.create_secret(
+                name="pull-secret",
+                namespace=target_ns,
+                data=pull_secret,
+                secret_type="kubernetes.io/dockerconfigjson",
+            )
+            # We are going to use the existence of the pull secret when
+            # we create the pod in order to tell whether we should patch
+            # in imagePullSecrets.  So we need to wait for it to show
+            # up before we move on to the pod creation part.
+            await self._wait_for_operation(
+                namespace=target_ns,
+                item="pull-secret",
+                operation=ObjectOperation.CREATION,
+                interval=0.2,
+            )
```

Comment:
Here too, do secrets not get created before the API call returns? I thought they did. (We were having problems before because we were using `VaultSecret`, which means the secret doesn't exist until vault-secrets-operator runs, but we don't have that problem in the new design.)

Reply:
Removed, and we just require that pull-secret be specified in order to patch it into the container's ImagePullSecrets.